Support bestline for line editing

MaxGyver83 opened the pull request:
# Support bestline for line editing

[bestline](https://github.com/jart/bestline) is a small line editing library (roughly 3000 lines of code) based on [linenoise](https://github.com/antirez/linenoise).

> Bestline is a fork of linenoise (a popular readline alternative) that fixes
> its bugs and adds the missing features while reducing binary footprint
> (surprisingly) by removing bloated dependencies, which means you can finally
> have a permissively-licensed command prompt w/ a 38kb footprint that's nearly
> as good as gnu readline.

## Installation

First, install a patched version of bestline as dynamic library:

```sh
git clone -b lib https://github.com/MaxGyver83/bestline.git
cd bestline
make libbestline.so
sudo cp libbestline.so /usr/local/lib/
sudo cp bestline.h /usr/local/include/
```

This patch adds a _position_ argument to bestline's `completion` function. The cursor position is necessary to decide which part of the current command is to be completed. The cursor position is updated for completion that don't happen at the end of a line. It also adds a `make` target for building a dynamic library. (In [integrate Antirez's linenoise single file zero conf line editing by rlonstein · Pull Request #5 · rakitzis/rc](https://github.com/rakitzis/rc/pull/5), it was suggested to provide linenoise as a standalone library.) I'll open a pull request with these changes to bestline.

Then, build rc with bestline support:

```sh
cd /path/to/rc
make EDIT=bestline
```

## Features

Bestline states to be "nearly as good as gnu readline".

Some differences compared to readline:

- No insertion of the last argument of the previous command (`Alt+.` in readline).
- Up arrow doesn't consider typed characters (`history-search-backward`). Use `Ctrl+r` instead.
- After deleting two words with `Ctrl+w w`, `Ctrl+y` inserts only the first word (with readline it inserts both words).
- Tab completion iterates over possible completions but doesn't print a list.

Bonus in this integration:

- Filename completion expands `~/dir` to `$HOME/dir`.

rakitzis commented:
Hi! Thanks for the contribution! In general, these edit libraries are "best effort" and I'm happy to add a new one. I'm curious if there is any automated testing that could be included with this PR. I'm happy to merge without, but I wonder how you tested the library integration.

MaxGyver83 commented:
Yes, I have written some unit tests. They depend on [mity/acutest: Simple header-only C/C++ unit testing facility](https://github.com/mity/acutest). I guess you don't want to include `acutest.h` in the `rc` repository. I could omit this file and leave it to future developers/testers to download it themselves. Or I could remove this dependency and use `assert`'s instead. (No report in this case.) In any case, I need to clean up the tests to make them work everywhere. (For example the command completion depends on the commands available on my laptop.)

MaxGyver83 commented:
I have fixed a bug in `edit-bestline.c` that showed up when I cleaned-up the unit tests.
And I have added the unit tests in a separate commit. They can be executed via `make run-test-bestline`. This will download the lastest version of `acutest.h`.
